onnxruntime/core/providers/qnn/builder/opbuilder/instance_norm_op_builder.cc

```diff
@@ -0,0 +1,136 @@
+
+  Status IsOpSupported(QnnModelWrapper& qnn_model_wrapper,
+                       const NodeUnit& node_unit,
+                       const logging::Logger& logger,
+                       bool is_quantized_model) const override final ORT_MUST_USE_RESULT;
+
+ protected:
+  Status ProcessAttributesAndOutputs(QnnModelWrapper& qnn_model_wrapper,
+                                     const NodeUnit& node_unit,
+                                     std::vector<std::string>&& input_names,
+                                     const logging::Logger& logger,
+                                     bool is_quantized_model,
+                                     bool do_op_validation) const override ORT_MUST_USE_RESULT;
+};
+
+// Instance normalization op is sensitive to data layout.
+// The nodes from 1st call of GetCapability do not get layout transformer applied, so their shapes are still NCHW.
+// The nodes from 2nd call of GetCapability get their layout transformed to NHWC.
+// Therefore, we need to check the node domain to determine if the layout has been transformed.
+Status InstanceNormOpBuilder::IsOpSupported(QnnModelWrapper& qnn_model_wrapper,
+                                            const NodeUnit& node_unit,
+                                            const logging::Logger& logger,
+                                            bool is_quantized_model) const {
+  ORT_UNUSED_PARAMETER(logger);
+
+  const auto float_elem_type = ONNX_NAMESPACE::Utils::DataTypeUtils::ToType("float");
+
+  // Check input type is float for CPU.
+  const auto& inputs = node_unit.Inputs();
+  ONNX_NAMESPACE::DataType input_data_type = inputs[0].node_arg.Type();
+  if (!is_quantized_model && input_data_type != float_elem_type) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm data type " + *input_data_type + " is not supported in CPU backend.");
+  }
+
+  // Also check output type is float for CPU.
+  const auto& outputs = node_unit.Outputs();
+  ONNX_NAMESPACE::DataType output_data_type = outputs[0].node_arg.Type();
+  if (!is_quantized_model && output_data_type != float_elem_type) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm data type " + *output_data_type + " is not supported in CPU backend.");
+  }
+
+  std::vector<uint32_t> input_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[0].node_arg, input_shape), "Cannot get shape of input 0");
+  const size_t input_rank = input_shape.size();
+
+  if (input_rank <= 2 || input_rank > 4) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm only supports input ranks of size 3 or 4.");
+  }
+
+  if (node_unit.Domain() != kMSInternalNHWCDomain) {  // Layout has not been transformed to NHWC, so do it here.
+    const uint32_t chans = input_shape[1];
+    for (size_t i = 1; i < input_rank - 1; ++i) {
+      input_shape[i] = input_shape[i + 1];
+    }
+
+    input_shape[input_rank - 1] = chans;
+  }
```

Comment:
why not 
if (node_unit.Domain() != kMSInternalNHWCDomain) {
   num_channels =input_shape[1];
}

I'm also wondering does it worth to do same thing again if node domain is kMSInternalNHWCDomain? It's useless to fallback nodes during the 2nd round of graph partitioning, because most of the case CPU EP can't support this node with layout transformed.

Reply:
> if (node_unit.Domain() != kMSInternalNHWCDomain) {
num_channels =input_shape[1];
}

I'll change it.

> I'm also wondering does it worth to do same thing again if node domain is kMSInternalNHWCDomain? It's useless to fallback nodes during the 2nd round of graph partitioning, because most of the case CPU EP can't support this node with layout transformed.

Are you suggesting that we return early from this function if we detect that the domain is `kMSInternalNHWCDomain`?

Reply:
We can leave it for now. It's common for all the layout sensitive ops in our op builder.